<a href="https://colab.research.google.com/github/trevorwjames/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Trevor_James_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [6]:
# Read in data set
# set column headers

import pandas as pd
import numpy as np

column_names = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 
                'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base',
                'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 
                'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower',
                'peak-rpm', 'city-mpg', 'higway-mpg', 'price']           
    

df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names= column_names)

df.head()

symboling normalized-losses         make  ... city-mpg higway-mpg  price
0          3                 ?  alfa-romero  ...       21         27  13495
1          3                 ?  alfa-romero  ...       21         27  16500
2          1                 ?  alfa-romero  ...       19         26  16500
3          2               164         audi  ...       24         30  13950
4          2               164         audi  ...       18         22  17450

[5 rows x 26 columns]

In [8]:
#replace ? values with NaN

df = df.replace({'?' : np.nan})

df.head()

symboling normalized-losses         make  ... city-mpg higway-mpg  price
0          3               NaN  alfa-romero  ...       21         27  13495
1          3               NaN  alfa-romero  ...       21         27  16500
2          1               NaN  alfa-romero  ...       19         26  16500
3          2               164         audi  ...       24         30  13950
4          2               164         audi  ...       18         22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [15]:
# create a 10 instance random sample
# view mean values for 'highway-mpg' for entire df and sample df
dfsample = df.sample(10, random_state=30)

print(df['higway-mpg'].mean())
print(dfsample['higway-mpg'].mean())

30.75121951219512
32.8


The two means are different, although we need to test the tvalue and pvalue in order to get a statistical difference

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [20]:
from scipy import stats

#testing the 1 sample below

stats.ttest_1samp(dfsample['higway-mpg'], 35.0)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

With the above sample size resulting in a -0.96 tvalue, and 36% pvalue, the confidence level is not adequate enough to reject the null hypothesis. 

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [29]:
# create a 100 car sample DF

dfsample100 = df.sample(100, random_state=30)

dfsample100.head()

symboling normalized-losses           make  ... city-mpg higway-mpg  price
121          1               154       plymouth  ...       31         38   6692
176         -1                65         toyota  ...       27         32  10898
24           1               148          dodge  ...       31         38   6229
88          -1               137     mitsubishi  ...       23         30   9279
72           3               142  mercedes-benz  ...       16         18  35056

[5 rows x 26 columns]

In [27]:
stats.ttest_1samp(dfsample100['higway-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

The results of testing a larger sample size, resulting in a tvalue of -6.52 and pvalue of .000000003 gives us sufficient confidence to reject the null hypthosis and suggest that the care salesmen is indeed a salesman and inflating the average Highway-mpg of the cars he sells. We would infer from our statistical finding that it would be lower than 35

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Because cars can have such different make-ups to result in largerly different mpg. A small sample size can include certain cars that drive the mean in a direction that may not be representative of the entire population. 

When taking a larger sample size we have the ability to average out those values and get a better indication of the actual mean value. 


## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [31]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-06-08 20:32:20--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-06-08 20:32:21 (137 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [32]:
#read in data
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers)

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [47]:
df = df.replace({'y':1, 'n':0, '?':np.nan})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
dem = df[df['party'] == 'democrat']
rep = df[df['party'] == 'republican']


### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

1. Null Hythothesis: The democrats support the budget bill **more** than the republicans

Alt Hypothesis: The democrats **DO NOT** support the budget bill more than the republicans.

In [41]:
stats.ttest_ind(dem['budget'], rep['budget'], nan_policy='omit')



Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

In [42]:
print(dem['budget'].mean())
print(rep['budget'].mean())

0.8884615384615384
0.13414634146341464


Result: Fail to Reject Null Hypothesis: More democrats support whatever they were voting on for the budget, than republicans. With the tvalue at 23.2, and the pvalue at 2.02^-77 it indicated, without a shadow of a doubt the democrats supported it. 

2.
Null Hypothesis: Republicans support the Physician Fee Freeze more than the Democrats

Alt Hypothesis: Republicans do not support the Physician Fee Freeze more than the democrats

In [39]:
stats.ttest_ind(rep['physician-fee-freeze'], dem['physician-fee-freeze'], nan_policy="omit")

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

In [43]:
print(rep['physician-fee-freeze'].mean())
print(dem['physician-fee-freeze'].mean())

0.9878787878787879
0.05405405405405406


Result: Fail to Reject Null Hypothesis: with a tvalue of 49.37 and pvalue of
1.99^-177 it seems the republicans were almost unaninmously for the physcian freeze and the democrats were entirely against it. 

3. 

Null Hyothesis: Thers is no significant differnce between the Repeublicans and Democrats in regards to voting for the water project

Alt Hypothesis: There is a significant difference between the Republicans and democrats in regards to voting for the water project

In [44]:
stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy="omit")

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [45]:
print(rep['water-project'].mean())
print(dem['water-project'].mean())

0.5067567567567568
0.502092050209205


Result: Fail to Reject Null Hypothesis: A very small tvalue, .089, and a relatively large pvalue, .93, leads me to believe that there was almost no difference in the way republicans abd democrats voted in regards to the water project

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [54]:
# create for loop to run ttests for all issues in the table

for column in column_headers:
    print(stats.ttest_ind(rep[column], dem[column], nan_policy='omit'))

TypeError: ignored



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





Null Hypothesis: More than 50% of republicans voted for the El Salvador Aid

Atl Hypothesis: Less than 50% of republicans voted for the El Salvador Aid

In [90]:
print(rep['el-salvador-aid'][0:11].mean())
print(rep['el-salvador-aid'].mean())

1.0
0.9515151515151515


In [94]:
repsample.describe()

count    11.0
mean      1.0
std       0.0
min       1.0
25%       1.0
50%       1.0
75%       1.0
max       1.0
Name: el-salvador-aid, dtype: float64

In [92]:
import numpy as np

((rep['el-salvador-aid'][0:11].mean()) - (rep['el-salvador-aid'].mean()))/(.215442/np.sqrt(10))

0.7116651034630435

In [85]:
repsample = rep['el-salvador-aid'][0:11].dropna()

repsample.head()

0     1.0
1     1.0
7     1.0
8     1.0
10    1.0
Name: el-salvador-aid, dtype: float64

In [86]:
stats.ttest_1samp(repsample, .5)

Ttest_1sampResult(statistic=inf, pvalue=0.0)

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE